composer/callbacks/run_directory_uploader.py

```diff
@@ -0,0 +1,250 @@
+          ensure that the GIL is not blocking the training loop when performance CPU
+          operations on uploaded files (e.g. comparing and computing checksums).
+          Network I/O happens always occurs in the background.
+
+        * Provide a RAM disk path for the `upload_staging_folder` parameter. Copying files to stage on RAM
+          will be faster than writing to disk. However, you must have sufficient excess RAM on your system,
+          or you may experience OutOfMemory errors.
+
+    .. note::
+
+        To use this callback, install composer with `pip install mosaicml[logging]`.
+
+    Args:
+        provider (str): Cloud provider to use.
+
+            Specify the last part of the Apache Libcloud Module here.
+            `This document <https://libcloud.readthedocs.io/en/stable/storage/supported_providers.html#provider-matrix>`
+            lists all supported providers. For example, the module name for Amazon S3 is `libcloud.storage.drivers.s3`, so
+            to use S3, specify 's3' here.
+
+        container (str): The name of the container (i.e. bucket) to use.
+        num_concurrent_uploads (int, optional): Maximum number of concurrent uploads. Defaults to 4.
+        upload_staging_folder (Optional[str], optional): A folder to use for staging uploads.
+            If not specified, defaults to using a :class:`~tempfile.TemporaryDirectory`.
+        use_procs (bool, optional): Whether to perform file uploads in background processes (as opposed to threads).
+            Defaults to True.
+        upload_every_n_batches (int, optional): Interval at which to scan the run directory for changes and to
+            queue uploads of files. Uploads are always queued at the end of the epoch. Defaults to every 100 batches.
+        provider_init_kwargs (Dict[str, Any], optional): Parameters to pass into the constructor for the
+            :class:`~libcloud.storage.providers.Provider` constructor. These arguments would usually include the cloud region
+            and credentials. Defaults to None, which is equivalent to an empty dictionary.
+    """
+
+    def __init__(
+        self,
+        provider: str,
+        container: Optional[str] = None,
+        num_concurrent_uploads: int = 4,
+        upload_staging_folder: Optional[str] = None,
+        use_procs: bool = True,
+        upload_every_n_batches: int = 100,
+        provider_init_kwargs: Dict[str, Any] = None,
+    ) -> None:
+        run_directory = get_run_directory()
+        if run_directory is None:
+            warnings.warn("NoRunDirectory: The run directory is not set, so the RunDirectoryUploader will be a no-op")
+            return
+
+        if provider_init_kwargs is None:
+            provider_init_kwargs = {}
+        self._provider_init_kwargs = provider_init_kwargs
+        self._upload_every_n_batches = upload_every_n_batches
+        self._object_name_prefix = ""  # TODO ravi. Decide how this will be set. Hparams? Run directory name?
```

Comment:
Perhaps everything related to the file upload config should just be in a config file that gets read to populate these fields.

Thinking about it from a user perspective, we'd like to avoid people having to hardcode any kind of sensitive info (including blobstore paths)  in their code

Reply:
It's fine to hardcode in everything except the secret. Blobstore paths should not be sensitive. I think it'd be common to hardcode a bucket to use, for example.